pkg/modifycontroller/modify_status.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package modifycontroller
+
+import (
+	"fmt"
+
+	"github.com/kubernetes-csi/external-resizer/pkg/util"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+// markControllerModifyVolumeStatus will mark ModifyVolumeStatus other than completed in the PVC
+func (ctrl *modifyController) markControllerModifyVolumeStatus(
+	pvc *v1.PersistentVolumeClaim,
+	modifyVolumeStatus v1.PersistentVolumeClaimModifyVolumeStatus,
+	err error) (*v1.PersistentVolumeClaim, error) {
+
+	newPVC := pvc.DeepCopy()
+	if newPVC.Status.ModifyVolumeStatus == nil {
+		newPVC.Status.ModifyVolumeStatus = &v1.ModifyVolumeStatus{}
+	}
+	newPVC.Status.ModifyVolumeStatus.Status = modifyVolumeStatus
+	// Update PVC's Condition to indicate modification
+	pvcCondition := v1.PersistentVolumeClaimCondition{
+		Type:               v1.PersistentVolumeClaimVolumeModifyingVolume,
+		Status:             v1.ConditionTrue,
+		LastTransitionTime: metav1.Now(),
+	}
+	conditionMessage := ""
+	switch modifyVolumeStatus {
+	case v1.PersistentVolumeClaimModifyVolumeInProgress:
+		conditionMessage = "ModifyVolume operation in progress."
+		newPVC.Status.ModifyVolumeStatus.TargetVolumeAttributesClassName = *pvc.Spec.VolumeAttributesClassName
+	case v1.PersistentVolumeClaimModifyVolumeInfeasible:
+		conditionMessage = "ModifyVolume failed with error" + err.Error() + ". Waiting for retry."
+	}
+	pvcCondition.Message = conditionMessage
+	if modifyVolumeStatus != v1.PersistentVolumeClaimModifyVolumePending {
+		newPVC.Status.Conditions = util.MergeModifyVolumeConditionsOfPVC(newPVC.Status.Conditions,
```

Comment:
It took me a moment to understand why we are skipping adding conditions if volume's modifiaction was pending. As I understand it, this is because - we do not want to change condition for penging modifications and keep older condition when this happens? Can we add a comment?

Reply:
We can add more comments, but end of the day, if we want to share functions(like the next comment about clean up) we will have more diverge in the code structure with this diagram: https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/3751-volume-attributes-class/VolumeAttributesClass-ModifyVolume-Flow-v3.png

And it is probably still confusing. I do not know what is the best way to do this.